main/src/ca/uwaterloo/flix/language/phase/MonomorphEnums.scala

```diff
@@ -0,0 +1,449 @@
+        // The enum is parametric on its type parameters, so we must instantiate
+        // e.g. for List[a] and List[Int32] we substitute [a -> Int32]
+        val e = root.enums(sym)
+        val subst = Substitution(e.tparams.map(_.sym).zip(args).toMap)
+        val cases = e.cases.map {
+          case (caseSym, LoweredAst.Case(_, caseTpe, caseSc, caseLoc)) =>
+            val freshCaseSym = new Symbol.CaseSym(freshSym, caseSym.name, caseSym.loc)
+            val newCaseSc = Scheme(caseSc.quantifiers, caseSc.tconstrs, caseSc.econstrs, subst(caseSc.base))
+            val caze = LoweredAst.Case(freshCaseSym, visitType(subst(caseTpe)), visitScheme(newCaseSc), caseLoc)
+            (freshCaseSym, caze)
+        }
+        val freshEnum = LoweredAst.Enum(
+          e.doc,
+          e.ann,
+          e.mod,
+          freshSym,
+          Nil,
+          e.derives,
+          cases,
+          Type.mkEnum(freshSym, Nil, loc),
+          e.loc
+        )
+        ctx.specializedEnums.put(freshSym, freshEnum)
+        freshSym
+    }
+  }
+
+  /**
+    * Returns a type where
+    * - The assumptions still hold
+    * - formulas in types have been evaluated (and ordered in the case of sets)
+    * - types involving rows have been sorted alphabetically (respecting duplicate label ordering)
+    *
+    * Assumes that
+    * - tpe is ground (no type variables
+    * - tpe has no aliases
+    * - tpe has no associated types
+    */
+  private def normalizeType(tpe: Type): Type = tpe match {
+    case Type.Var(sym, loc) => throw InternalCompilerException(s"Unexpected type var '$sym'", loc)
+    case Type.Cst(_, _) => tpe
+    case Type.Apply(tpe1, tpe2, applyLoc) =>
+      val t1 = normalizeType(tpe1)
+      val t2 = normalizeType(tpe2)
+      t1 match {
+        // Simplify effect set equations.
+        case Type.Cst(TypeConstructor.Complement, _) => t2 match {
+          case Type.Pure => Type.EffUniv
+          case Type.EffUniv => Type.Pure
+          case _ => throw InternalCompilerException(s"Unexpected non-simple effect $tpe", applyLoc)
+        }
+        case Type.Apply(Type.Cst(TypeConstructor.Union, _), x, _) =>
+          (x, t2) match {
+            case (Type.Pure, Type.Pure) => Type.Pure
+            case (Type.Pure, Type.EffUniv) => Type.EffUniv
+            case (Type.EffUniv, Type.Pure) => Type.EffUniv
+            case (Type.EffUniv, Type.EffUniv) => Type.EffUniv
+            case _ => throw InternalCompilerException(s"Unexpected non-simple effect $tpe", applyLoc)
+          }
+        case Type.Apply(Type.Cst(TypeConstructor.Intersection, _), x, _) =>
+          (x, t2) match {
+            case (Type.Pure, Type.Pure) => Type.Pure
+            case (Type.Pure, Type.EffUniv) => Type.Pure
+            case (Type.EffUniv, Type.Pure) => Type.Pure
+            case (Type.EffUniv, Type.EffUniv) => Type.EffUniv
+            case _ => throw InternalCompilerException(s"Unexpected non-simple effect $tpe", applyLoc)
+          }
```

Comment:
Is the above necessary? Smart constructors maybe? or?

Reply:
Smart constructors do not guarantee that formulas are completely evaluated, even the type minimizer does not guarantee it

Reply:
Hmm. @mlutze Thoughts?

It seems we do simplications in so many places :O

Reply:
I like that this is separate from other type minimization due to the guarantees, as Jonathan said. Maybe later we can refactor if we find places that "lawful normalization" should be used.